lib/shanty_docker_plugin/docker_plugin.rb

```diff
@@ -0,0 +1,100 @@
+require 'shanty/plugin'
+require 'shanty/project'
+require 'shanty_docker_plugin/docker_credentials'
+require 'shanty_docker_plugin/image_resolution_fallback'
+require 'rugged'
+
+module DockerPlugin
+  # Shanty docker plugin
+  class DockerPlugin < Shanty::Plugin
+    NAME_OPTION = :docker_project_name
+    DEPENDENCY_OPTION = :docker_dependency
+
+    def initialize(*args)
+      super(*args)
+      @docker_crendentials = DockerCredentials.new
+    end
+
+    provides_config NAME_OPTION
+    provides_config :registry
+    provides_config :tag
+    provides_config :output_dir, 'build'
+    provides_config :fallback_tag, 'latest'
+    subscribe :build, :on_build
+    provides_projects :resolve_projects
+    description 'Discovers and builds Docker projects'
+
+    def self.resolve_projects(env)
+      project_dependencies(find_projects(env))
+    end
+
+    def on_build
+      if project.changed?
+        build
+      else
+        @docker_crendentials.auth!(image_wrapper.registry)
+        ImageResolutionFallback.pull!(image_wrapper, config[:fallbacktag])
+      end
+    end
+
+    def on_deploy
+      @docker_crendentials.auth!(image_wrapper.registry)
+      image_wrapper.underlying.push
+    end
+
+    private
+
+    def self.find_projects(env)
```

Comment:
All of these `self.whatever` methods are not affected by `private` - an annoying quirk in Ruby. Instead, you need to call `private_class_method` to make them private. See [this blog post for more information](http://domon.cc/2013/12/25/private-class-methods-in-ruby/).